lib/mappers/esdn_mapper.py

```diff
@@ -0,0 +1,348 @@
+                        mapped_props["object"] = url.get("#text")
+                except Exception as e:
+                    logger.error("Error getting mapping isShownAt and object" +
+                                 " %s\nException:\n%s" %
+                                 (self.provider_data["_id"], e))
+
+            self.mapped_data.update(self.clean_dict(mapped_props))
+
+    def map_language(self):
+        prop = self.root_key + "language/languageTerm"
+        lang = None
+        if exists(self.provider_data, prop):
+            lang = getprop(self.provider_data, prop)
+            if isinstance(lang, dict):
+                lang = lang.get("#text")
+
+        if lang:
+            self.update_source_resource({"language": lang})
+
+    def map_relation(self):
+        prop = self.root_key + "relatedItem/titleInfo/title"
+
+        if exists(self.provider_data, prop):
+            self.update_source_resource({
+                "relation":
+                    getprop(self.provider_data,
+                            prop)
+            })
+
+    def map_rights(self):
+        prop = self.root_key + "accessCondition"
+        rights = []
+        if exists(self.provider_data, prop):
+            for right in iterify(getprop(self.provider_data, prop)):
+                if isinstance(right, dict):
+                    rights.append(right.get("#text"))
+                elif isinstance(right, dict):
+                    for r in right:
+                        rights.append(r)
+                else:
+                    rights.append(right)
+        if rights:
+            self.update_source_resource({"rights": rights})
+
+    def map_subject_and_spatial_and_temporal(self):
+        prop = self.root_key + "subject"
+        mapped_props = {
+            "subject": [],
+            "spatial": [],
+            "temporal": []
+        }
+
+        if exists(self.provider_data, prop):
+            for s in iterify(getprop(self.provider_data, prop)):
+                try:
+                    if "geographic" in s and s.get("geographic"):
+                        mapped_props["spatial"].append(s.get("geographic"))
+                    elif "topic" in s and s.get("topic"):
+                        mapped_props["subject"].append(s.get("topic"))
+                    elif "temporal" in s and s.get("temporal"):
+                        mapped_props["temporal"].append(s.get("temporal"))
+                except Exception as e:
+                    logger.error("Error mapping geo/subject/temporal"
+                                 "for record %s\nException:\n%s" %
+                                 (self.provider_data["_id"], e))
+
+            for k in mapped_props.keys():
+                mapped_props[k] = self.unnest_list(mapped_props.get(k), [])
+
+            if mapped_props:
+                self.update_source_resource(self.clean_dict(
+                    mapped_props))
+
+    def map_title(self):
```

Comment:
Regarding the rest of this function, the test for `s.keys()` still has me wondering. Was this function copied from somewhere else in `ingestion`? If so, it could simply be wrong -- from an earlier period in which there were no code reviews. :-)

The safest thing is probably to have one condition (with no `or` clause) and check for `isinstance(s, basestring)` and have that be it. In Python 2.7, strings default to the ASCII encoding, and it would be an assumption -- perhaps a safe assumption, but an assumption -- that we're being provided with `unicode` by the XML processing library or whatever. The success of this section of code shouldn't depend on whether it's unicode or any other character set, just that it's a string.